README
README for Clutter 1.10.6
===============================================================================

Clutter is an open source software library for creating fast, compelling,
portable, and dynamic graphical user interfaces.

REQUIREMENTS
-------------------------------------------------------------------------------

Clutter currently requires:

  • GLib ≥ 2.31.19
  • Cogl ≥ 1.9.6
  • JSON-GLib ≥ 0.12.0
  • Atk ≥ 2.1.5
  • Cairo ≥ 1.10
  • PangoCairo ≥ 1.20

When building the X11 backend, Clutter depends on the following extensions:

  • XComposite ≥ 0.4
  • XDamage
  • XExt
  • XFixes ≥ 3
  • XInput (1.x or 2.x)
  • XKB

When building the Wayland backend, Clutter also depends on:

  • wayland-client
  • xkbcommon

When building the GDK backend, Clutter also depends on:

  • gdk-3.0 > 3.3.18

When building the CEx100 backend, Clutter also depends on:

  • libgdl

If you are building the API reference you will also need:

  • GTK-Doc ≥ 1.15

If you are building the additional documentation you will also need:

  • xsltproc
  • jw (optional, for generating PDFs)

If you are building the Introspection data you will also need:

  • GObject-Introspection ≥ 0.9.5

GObject-Introspection is available from:

  git://git.gnome.org/gobject-introspection

If you want support for profiling Clutter you will also need:

  • UProf ≥ 0.3

UProf is available from:

  git://github.com/rib/UProf.git

RESOURCES
-------------------------------------------------------------------------------

The official Clutter website is:

   http://www.clutter-project.org/

The API references for the latest stable release are available at:

   http://docs.clutter-project.org/docs/clutter/stable/
   http://docs.clutter-project.org/docs/cogl/stable/
   http://docs.clutter-project.org/docs/cally/stable/

The Clutter Cookbook is available at:

   http://docs.clutter-project.org/docs/clutter-cookbook/1.0/

New releases of Clutter are available at:

   http://source.clutter-project.org/sources/clutter/
   http://download.gnome.org/sources/clutter/

The Clutter blog is available at:

   http://www.clutter-project.org/blog/

To subscribe to the Clutter mailing lists and read the archives, use the
Mailman web interface available at:

   http://lists.clutter-project.org/

New bug page on Bugzilla:

   http://bugzilla.gnome.org/enter_bug.cgi?product=clutter

Clutter is licensed under the terms of the GNU Lesser General Public
License, version 2.1 or (at your option) later.

BUILDING AND INSTALLATION
-------------------------------------------------------------------------------

To build Clutter from a release tarball, the usual autotool triad should
be followed:

  $ ./configure
  $ make
  # make install

To build Clutter from a Git clone, run the autogen.sh script instead
of the configure one. The autogen.sh script will run the configure script
for you, unless the NOCONFIGURE environment variable is set to a non-empty
value.

See also the wiki page:

  http://wiki.clutter-project.org/wiki/BuildingClutter

Clutter has additional command line options for the configure script:

 --enable-debug=[no/minimum/yes]
	Controls Clutter debugging level:

	yes:
                All GLib asserts, checks and support for runtime Clutter
                debugging notes through CLUTTER_DEBUG. This is the default
                value for developers snapshots.

        minimum:
                Just GType cast checks and support for runtime Clutter
                debugging notes through CLUTTER_DEBUG. This is the default
                for stable releases.

        no:
                No GLib asserts or checks and no support for runtime Clutter
                debugging notes. Only use in extreme performance and/or size
                optimization cases, though it is strongly discouraged.

 --enable-maintainer-flags=[no/yes/error]
        Use strict compiler flags. This defaults to 'yes' for building from
        Git to 'no' for tarball releases. If 'error' is used, then -Werror
        will be enabled (if available).

 --enable-gtk-doc
	use gtk-doc to build API documentation (default=no). Requires gtk-doc
	present on the target system.

 --enable-docs=[no/yes]
        Build additional documentation. Requires xsltproc for DocBook
        conversion, and optionally jw for PDF generation.

 --enable-gcov=[no/yes]
        Build Clutter with coverage report support, provided by gcov. This
        feature only works with the GNU Compiler Suite and gcov installed.

 --enable-profile=[no/yes]
        Build Clutter with profiling instrumentation. Requires the GNU
        C Compiler and the UProf library.

 --enable-conform=[yes/no]
        Build the Clutter conformance test suite.

 --enable-deprecated=[yes/no]
        Whether deprecated symbols should be available when compiling Clutter.

 --disable-Bsymbolic
        Disable linking with -Bsymbolic.

 --enable-x11-backend=[yes/no/check]
        Enable the X11 backend. (default=check)

 --enable-win32-backend=[yes/no/check]
        Enable the native Microsoft Windows backend. (default=check)

 --enable-quartz-backend=[yes/no/check]
        Enable the native Quartz backend. (default=check)

 --enable-gdk-backend=[yes/no/check]
        Enable the GDK backend. (default=check)

 --enable-wayland-backend=[yes/no]
        Enable the Wayland client backend. (default=no) [EXPERIMENTAL]

 --enable-cex100-backend=[yes/no]
        Enable the CEx100 platform backend. (default=no) [EXPERIMENTAL]

 --enable-egl-backend=[yes/no]
        Enable the EGL framebuffer backend. (default=no)

 --enable-tslib-input=[yes/no]
        Enable the TSLib input backend. (default=no) [EXPERIMENTAL]

 --enable-evdev-input=[yes/no]
        Enable the evdev input backend. (default=no) [EXPERIMENTAL]

See also the INSTALL file generated by autotools for further information.

VERSIONING
-------------------------------------------------------------------------------

Clutter uses the common "Linux kernel" versioning system, where
even-numbered minor versions are stable and odd-numbered minor
versions are development snapshots.

Different major versions break both API and ABI but are parallel
installable. The same major version with differing minor version is
expected to be ABI compatible with other minor versions; differing
micro versions are meant just for bug fixing. On odd minor versions
the newly added API might still change.

The micro version indicates the origin of the release: even micro
numbers are only used for released archives; odd micro numbers are
only used on the Git repository.

HACKING
-------------------------------------------------------------------------------

If you want to hack on and improve Clutter check the HACKING file for
general implementation guidelines, and the HACKING.backends for
backend-specific implementation issues.

The CODING_STYLE file contains the rules for writing code conformant to the
style guidelines used throughout Clutter. Remember: the coding style is
mandatory; patches not conforming to it will be rejected by default.

BUGS
-------------------------------------------------------------------------------

Bugs should be reported to the Clutter Bugzilla at:

  http://bugzilla.gnome.org/enter_bug.cgi?product=clutter

You will need a Bugzilla account.

In the report you should include:

  • what system you're running Clutter on;
  • which version of Clutter you are using;
  • which version of GLib and OpenGL (or OpenGL ES) you are using;
  • which video card and which drivers you are using, including output of
    glxinfo and xdpyinfo (if applicable);
  • how to reproduce the bug.

If you cannot reproduce the bug with one of the tests that come with Clutter
source code, you should include a small test case displaying the bad
behaviour.

If the bug exposes a crash, the exact text printed out and a stack trace
obtained using gdb are greatly appreciated.

CONTRIBUTING
-------------------------------------------------------------------------------

Patches should be submitted using Bugzilla. Patches fixing a bug should be
attached to the bug report; patches for new features or for fixing bugs not
yet reported should be attached to a newly opened bug.

Patches should always be in the unified diff format, using:

  diff -Nuarp clutter.source clutter.patched > clutter-patch.diff

If diffing against the Git repository, you should use:

  git diff > clutter-patch.diff

Or, better: commit locally and use `git format-patch` to generate a patch
containing authorship details, so that members of the Clutter development
team can credit your contribution properly.

Another useful tool for interacting with Git and Bugzilla is git-bz(1):

  http://git.fishsoup.net/man/git-bz.html

Which is available here:

  http://git.fishsoup.net/cgit/git-bz/

RELEASE NOTES
-------------------------------------------------------------------------------

Relevant information for developers with existing Clutter applications
wanting to port to newer releases (see NEWS for general information on new
features).

Release Notes for Clutter 1.10
-------------------------------------------------------------------------------

• The ClutterActor::paint, ClutterActor::queue-redraw, and
  ClutterActor::queue-relayout signals are now annotated as not allowing
  signal emission hooks; this enables some optimization inside GLib, and
  given the amount of emissions these three signals have, we want to get
  through all the fast paths we can.

• The ClutterActorBox parameter of the ClutterActor::allocation-changed
  signal is now marked as G_SIGNAL_TYPE_STATIC_SCOPE.

• The clutter-stage-window.h header is not installed any more; this header
  was never meant to be public in the first place, and nothing could have
  been implementing the ClutterStageWindow interface and use it with Clutter.

• ClutterActor will call its unmap() implementation when it is removed from
  its parent. This may happen after the ClutterActor::destroy signal has been
  emitted, i.e. during its dispose(). Prior to 1.10, calling the
  clutter_actor_destroy() function would unmap the actor first, and then
  emit the destroy signal. This means that extra care should be used when
  overriding the map() and unmap() virtual functions, to avoid dereferencing
  NULL pointers. It is also worthy of note that since Clutter 1.8, overriding
  map() or unmap() is not required any more for subclasses of ClutterActor
  that have children, as ClutterActor will automatically do the right thing
  inside its own implementation, and map or unmap its children when needed.

• ClutterBox and ClutterGroup have been deprecated. ClutterActor should be
  used directly, instead. Since ClutterStage inherits from ClutterGroup, the
  instance and class structures are still available, but subclassing Group
  is strongly disencouraged.

• ClutterContainer provides default implementations for add(), remove(),
  foreach(), raise(), lower(), and sort_depth_order(); this means that
  overriding these virtual functions is now deprecated, and developers
  should only implement ClutterContainer for classes providing child
  properties.

• ClutterActor is not an abstract type any more, and can be instantiated
  either through clutter_actor_new() or through g_object_new().

• ClutterActor now implements ClutterContainer, and takes over the entire API
  for modifying the scene graph; this means that every actor can have children
  and it's not necessary any more to implement the ClutterContainer interface.
  ClutterActor provides new API to replace clutter_actor_set_parent() and
  clutter_actor_unparent(), as well as a default implementation of every
  Container virtual function. Existing actors overriding ClutterContainer and
  providing an alternate implementation will continue to work, even though it
  is strongly encouraged to port to the new API.

• The ClutterActor::destroy signal is going to be emitted at the beginning
  of the dispose sequence of a ClutterActor, instead of the end. This allows
  to access the state of the actor being destroyed, instead of just being
  able to use its pointer. This may expose bugs in code that does not check
  for NULL actor pointers.

• The long since broken depth cueing support in ClutterStage has been
  deprecated; the fixed pipeline fog API has been pretty much useless
  since the switch to colors and textures with premultiplied alpha
  channels. Setting ClutterStage:use-fog to TRUE has no visible results,
  and querying the :use-fog and :fog properties will always yield the
  default values.

• ClutterDeformEffect switched from using CoglVertexBuffer to using the
  CoglPrimitive API internally, to improve performance and use non-deprecated
  Cogl API. CoglPrimitive converts COGL_WRAP_MODE_AUTOMATIC to
  COGL_WRAP_MODE_CLAMP_TO_EDGE, unlike CoglVertexBuffer which converts it to
  COGL_WRAP_MODE_REPEAT. This prevents artifacts when sampling texture
  coordinates outside the [ 0, 1 ] range. This change may cause the back
  texture to not be painted if its coordinates go outside the allowed range,
  for instance when using a custom transformation matrix on the back material
  used by the ClutterDeformEffect.

• The "default stage" has been deprecated; since the 1.0 release, the default
  stage creation was deferred to the call to clutter_stage_get_default(), and
  the preferred way for getting a ClutterStage was calling clutter_stage_new()
  instead. On platforms that do not support multiple stages, attempting to
  create more than one stage will lead to an error, and Clutter will abort.

• Clutter can support multiple backends in the same shared library. Only one
  windowing or input backend can be used at run time. As a result of this
  change, the shared library name used by Clutter has changed from:

        libclutter-<flavour>-<API version>.so

  to:

        libclutter-<API version>.so

  The pkg-config file has been updated accordingly. Until the next major API
  break, Clutter will ship compatibility links for all the previous "flavours"
  that were available in versions < 1.10; this allows applications dynamically
  linking against Clutter, or using dlopen(), to keep working. For libraries
  and applications dynamically linking against Clutter, though, it is still
  recommended to recompile to make sure that the most recent version is being
  used. Language bindings using GObject Introspection will automatically use
  the new shared library without requiring any change.

• The windowing system backend for the CE3100 and CE4100 platforms using the
  libgdl library is now implemented as a separate backend instance, instead
  of being a sub-flavour of the EGL native framebuffer backend. This change
  introduces a new header file, under $includedir/clutter-1.0/clutter/cex100,
  which should be included to access the CEx100-specific API. The API and
  ABI of the platform API has not been changed, though it should still be
  considered experimental.

• As of 1.10 it is not necessary any more to call clutter_threads_init() to
  initialize threading support in Clutter; after the changes in GLib 2.32,
  threading support in Clutter is always enabled. The rules on how to use
  Clutter from multiple threads haven't changed.

• Deprecated API is now marked using the CLUTTER_DEPRECATED and the
  CLUTTER_DEPRECATED_FOR annotations; these two annotations will result in
  compiler warnings when attempting to use the deprecated API. It is possible
  to disable deprecation warnings for Clutter by defining the
  CLUTTER_DISABLE_DEPRECATION_WARNINGS symbol when compiling. The previous
  deprecation symbol, CLUTTER_DISABLE_DEPRECATED, is only used for macros.

• Deprecated functionality has been moved to separate header files, installed
  under the $includedir/clutter-1.0/clutter/deprecated directory. These files
  are still included by default by clutter/clutter.h.

Release Notes for Clutter 1.8
-------------------------------------------------------------------------------

• Cogl has been split out of tree. Clutter depends on Cogl, so users should
  not notice anything. Developers using pkg-config will gain an extra Cogl
  dependency.

• The clutter_actor_get_gid() and clutter_get_actor_by_gid() functions have
  been deprecated. They should not have been public in the first place.

• ClutterShader and its relative API have been deprecated. ClutterShaderEffect
  and the effect API is the proper way to use GLSL shaders with ClutterActors.
  It is possible to instantiate a ClutterShaderEffect directly.

• ClutterText's paint volume has been implemented using the ink rectangle of
  the PangoLayout used to paint, to allow culling to work properly even in
  the case of glyphs spilling outside the logical rectangle.

• Non fully opaque children of a non fully opaque actor will be composited
  in an offscreen framebuffer object to preserve the correct overall opacity.
  This changes the way non fully opaque actors are drawn, and might cause
  some code relying on the old, unspecified behaviour to produce different
  results.

• ClutterBoxLayout layout algorithm has been changed to map the more mature
  and more stable GtkBox one.

• The ClutterText:editable property was incorrectly defined in the GParamSpec
  as being TRUE by default, but it was initialized to FALSE. To avoid breaking
  existing code, the default value inside the GParamSpec (and inside the
  documentation) has been changed to FALSE as well.

• The preferred way to draw inside a ClutterCairoTexture is connecting to
  its ::draw signal, and calling the invalidate() method. The old pattern
  of creating the Cairo context using the create() method is deprecated but
  still working as intended.

• ClutterScore has been deprecated; the preferred way to chain up multiple
  animations is to use the ClutterAnimator or ClutterState classes.

Release Notes for Clutter 1.6
-------------------------------------------------------------------------------

• The internal copy of JSON-GLib has been removed: Clutter now strictly
  depends on the installed copy of this library. The --with-json configure
  switch has been removed.

• The ClutterBehaviour class and its sub-classes have been deprecated; the
  Clutter API reference contains a migration guide to port code based on
  behaviours to the implicit animations framework.

• The ClutterTimeoutPool and clutter_frame_source_* API have been deprecated;
  both API are not integrated in the Clutter main loop, and are not used
  internally any longer, so they are of relative use.

• It it not necessary any more to provide implementations for the
  ClutterActor::map() and ClutterActor::unmap() virtual functions, even
  for composite actors not implementing the ClutterContainer interface.

• ClutterTimeline now guarantees that the ::new-frame signal will be
  emitted at the beginning of the timeline, as well as guaranteeing that
  the ::completed signal will be emitted at the end of the timeline.

• ClutterActor will check for the enabled property of ClutterActorMeta
  instances (actions, constraints and effects), and will not invoke
  ClutterActorMeta functions on disabled instances. This removes the
  requirement for checking the property inside the ClutterActorMeta
  sub-classes.

• ClutterActorBox, ClutterGeometry and ClutterVertex install a progress
  function for ClutterInterval, allowing the interpolation of properties
  using them as storage types.

• ClutterColor's clutter_color_from_string() function accepts CSS3 color
  specifications.

• The previously unused "axes" field in the ClutterButtonEvent,
  ClutterScrollEvent and ClutterMotionEvent structures is now used on
  the X11-based backends with XInput support.

• ClutterListModel will honour the filter function when calling
  clutter_model_get_iter_at_row().

• ClutterClickAction does not use a pointer grab any longer, and uses
  a capture on the stage instead.

• On all platforms which allow it, ClutterStage will ask the windowing
  system for an explicit key focus when showing the stage window. This
  can be disabled using clutter_stage_set_accept_focus().

Release Notes for Cogl 1.6
-------------------------------------------------------------------------------

• Cogl may internally optimise cogl_read_pixels when only a single
  pixel is drawn and the entire scene is comprised of solid colour
  rectangles. Instead of actually rendering the rectangles it will
  compute the single pixel value in software. This effectively means
  that Clutter can do software picking without any API changes.

• Internally Cogl now has a GLSL backend as well as the ARBfp and
  fixed function backends. By default, this backend has the lowest
  priority but it can be explicitly enabled by setting the
  COGL_DEBUG environment variable, e.g.:

    export COGL_DEBUG=disable-fixed,disable-arbfp

  for builds of Clutter with debug support enabled.

• Cogl will internally now cache generated ARBfp programs so that it
  should be able to reuse a previous program even if it was generated
  for an unrelated CoglMaterial. This makes using one-shot materials
  less expensive than before although it is still not recommended.

Release Notes for Clutter 1.4
-------------------------------------------------------------------------------

• ClutterLayoutManager sub-classes overriding the set_container() virtual
  function should chain up to the parent class's implementation.

• The ClutterTexture:filename property is now readable, as well as writable;
  this allows querying the Texture for the filename storing the image data.
  The value will be unset when using the set_from_*_data() family of
  functions.

• If both the :sync-size and the :keep-aspect-ratio properties of a
  ClutterTexture are set to TRUE, then the texture actor will update its
  ClutterActor:request-mode property depending on the orientation of the
  image data - height-for-width for landscape, and width-for-height for
  portrait. Square image data will default to height-for-width, like all
  actors. You can still explicitly override the :request-mode value, or

NEWS
Clutter 1.10.6                                                       2012-05-28
===============================================================================

  • List of changes since Clutter 1.10.4

    - Fix assertions and segfaults with touch events

    - Drop usage of GL types in Clutter
    C, GLib, and Cogl provide the same exact types, and we also get to drop
    the dependency on the GL header.

    - Update Visual Studio project files and build rules

    - Fix assertion error in ClutterFlowLayout

    - Re-allow passing NULL in clutter_text_set_text()
    The switch to ClutterTextBuffer introduced a regression.

    - Add missing annotations in ClutterText

  • List of bugs fixed since Clutter 1.10.4

    [bugzilla.gnome.org]
    #675371 - Asserts when getting touch events
    #676150 - Fix GDK backend build on Windows
    #676068 - Setting size on ClutterFlowLayout container crashes
    #675890 - Passing NULL buffer to clutter_text_set_text() will not reset
              the buffer and segfault
    #675396 - Possible bug in _clutter_paint_volume_axis_align

Many thanks to:

  Chun-wei Fan, Bastien Nocera, Bastian Winkler, Dominique Bureau,
  Mike Ruprecht, Neil Roberts, Sasi Bhushan, Tristan Van Berkom.

Clutter 1.10.4                                                       2012-05-01
===============================================================================

  • List of changes since Clutter 1.10.2

    - Fixes for XInput 2 event handling
    We should only select events for enabled devices, and keep track of the
    device state.

    - Fix an overflow when calculating the distance of very long paths

    - Ensure that we don't change actor visibility when restacking

  • List of bugs fixed since Clutter 1.10.2

    [bugzilla.gnome.org]
    #674510 - Changing a hidden child's position with new API shows the actor
    #673644 - Scroll not working in document overview
    #652521 - get_length fails in Clutter.Path when length is greater
              than 46340
Many thanks to:

  Cosimo Cecchi, Christian Kirbach, Jasper St. Pierre, Jonh Wendell,
  Yuri Myasoedov.

Clutter 1.10.2                                                       2012-04-16
===============================================================================

  • List of changes since Clutter 1.10.0

    - Fix expansion of actors inside a TableLayout
    For actors that span more than one column, the wrong flag was being
    used to compute the expansion.

    - Fix text length passed to ClutterTextBuffer
    This should fix crashers happening when changing the contents of a
    TextBuffer with non-ASCII characters.

    - Clean up handling of in-flight changes during implicit transitions
    The transitions are updated using the new value, as well as the
    current easing state. The documentation has been updated to define
    the behaviour of the implicit animations.

  • List of bugs fixed since Clutter 1.10.0

    [bugzilla.gnome.org]
    #673783 - ClutterTextBuffer crashes
    #672945 - Fixes for the implicit actor animation API
    #672853 - Assistant layout is broken
    #672890 - CRASH: ClutterAnimation finalized without disconnecting timeline
    #674105 - cally.pc is buggy

Many thanks to:

  Jasper St. Pierre, Neil Roberts, Owen W. Taylor, Peter Hutterer,
  Stefano Facchini, Tristan Van Berkom.

Clutter 1.10.0                                                       2012-03-22
===============================================================================

  • Clutter 1.10.0, new stable release

  • List of changes since Clutter 1.9.16

    - Fix ClutterTimeline:current-repeat accessor
    The :current-repeat value was being incremented prior to emitting the
    completed signal, which meant that signal handlers would not be able
    to retrieve the correct value.

    - Improve the memory management of Transitions
    ClutterActor now takes a reference on the Transitions that have been
    added to it, and releases the reference when the Transition is removed;
    this means that it's safe to release the reference of a transition after
    adding it, to let ClutterActor own it.

    - Remove the Animatable argument for PropertyTransition constructor
    The animatable was not meant to be there in the first place; the
    PropertyTransition will perform validation on the property lazily, once
    it's been attached to the actor.

    - Clean up documentation and examples
    Ensure that the API reference is using the correct API and best
    practices.

    - Fix TableLayout policy for actors spanning multiple rows or columns
    The ClutterTableLayout inherited some small bug from MxTable that got
    fixed in the meantime; the layout policy has been re-synchronized.

    - Build fixes for the Wayland backend.

    - Updated VisualStudio project files.

  • List of bugs fixed since Clutter 1.9.16

    [bugzilla.gnome.org]
    #672557 - Fix ClutterTableLayout

Many thanks to:

  Jasper St. Pierre, Chun-wei Fan, Robert Bragg, Chao-Hsiung Liao,
  Neil Roberts, Rico Tzschichholz, Rob Bradford

Clutter 1.9.16                                                       2012-03-19
===============================================================================

  • Clutter 1.10.0, release-candidate 1

  • List of changes since Clutter 1.9.14

    - Support implicit animatable properties
    Various properties of ClutterActor can now be implicitly animated when
    using the public setters, instead of going through the animation API;
    this is an opt-in mechanism for the 1.x API series.

    - Introduce paint nodes
    ClutterPaintNode is an element of the render tree; it is used to store
    the pipeline state and the geometry to be drawn. Actors have the ability
    to be painted using the paint nodes by overriding the paint_node() virtual
    function instead of the paint() virtual function. For the 1.x API series
    the render tree will be consumed per-actor, but the plan is to build a

clutter/clutter-version.h
/*
 * Clutter.
 *
 * An OpenGL based 'interactive canvas' library.
 *
 * Authored By Matthew Allum  <mallum@openedhand.com>
 *
 * Copyright (C) 2006 OpenedHand
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library. If not, see <http://www.gnu.org/licenses/>.
 *
 *
 */

#if !defined(__CLUTTER_H_INSIDE__) && !defined(CLUTTER_COMPILATION)
#error "Only <clutter/clutter.h> can be included directly."
#endif

#ifndef __CLUTTER_VERSION_H__
#define __CLUTTER_VERSION_H__

/**
 * SECTION:clutter-version
 * @short_description: Versioning utility macros
 *
 * Clutter offers a set of macros for checking the version of the library
 * at compile time; it also provides a function to perform the same check
 * at run time.
 *
 * Clutter adds version information to both API deprecations and additions;
 * by definining the macros %CLUTTER_VERSION_MIN_REQUIRED and
 * %CLUTTER_VERSION_MAX_ALLOWED, you can specify the range of Clutter versions
 * whose API you want to use. Functions that were deprecated before, or
 * introduced after, this range will trigger compiler warnings. For instance,
 * if we define the following symbols:
 *
 * |[
 *   CLUTTER_VERSION_MIN_REQUIRED = CLUTTER_VERSION_1_6
 *   CLUTTER_VERSION_MAX_ALLOWED  = CLUTTER_VERSION_1_8
 * ]|
 *
 * and we have the following functions annotated in the Clutter headers:
 *
 * |[
 *   void clutter_function_A (void) CLUTTER_DEPRECATED_IN_1_4;
 *   void clutter_function_B (void) CLUTTER_DEPRECATED_IN_1_6;
 *   void clutter_function_C (void) CLUTTER_AVAILABLE_IN_1_8;
 *   void clutter_function_D (void) CLUTTER_AVAILABLE_IN_1_10;
 * ]|
 *
 * then any application code using the functions above will get the output:
 *
 * |[
 *   clutter_function_A: deprecation warning
 *   clutter_function_B: no warning
 *   clutter_function_C: no warning
 *   clutter_function_D: symbol not available warning
 * ]|
 *
 * It is possible to disable the compiler warnings by defining the macro
 * %CLUTTER_DISABLE_DEPRECATION_WARNINGS before including the clutter.h
 * header.
 */

#include <glib.h>

G_BEGIN_DECLS

/**
 * CLUTTER_MAJOR_VERSION:
 *
 * The major version of the Clutter library (1, if %CLUTTER_VERSION is 1.2.3)
 */
#define CLUTTER_MAJOR_VERSION   (1)

/**
 * CLUTTER_MINOR_VERSION:
 *
 * The minor version of the Clutter library (2, if %CLUTTER_VERSION is 1.2.3)
 */
#define CLUTTER_MINOR_VERSION   (10)

/**
 * CLUTTER_MICRO_VERSION:
 *
 * The micro version of the Clutter library (3, if %CLUTTER_VERSION is 1.2.3)
 */
#define CLUTTER_MICRO_VERSION   (6)

/**
 * CLUTTER_VERSION:
 *
 * The full version of the Clutter library, like 1.2.3
 */
#define CLUTTER_VERSION         1.10.6

/**
 * CLUTTER_VERSION_S:
 *
 * The full version of the Clutter library, in string form (suited for
 * string concatenation)
 */
#define CLUTTER_VERSION_S       "1.10.6"

/**
 * CLUTTER_VERSION_HEX:
 *
 * Numerically encoded version of the Clutter library, like 0x010203
 */
#define CLUTTER_VERSION_HEX     ((CLUTTER_MAJOR_VERSION << 24) | \
                                 (CLUTTER_MINOR_VERSION << 16) | \
                                 (CLUTTER_MICRO_VERSION << 8))

/* XXX - Every new stable minor release bump should add a macro here */

/**
 * CLUTTER_VERSION_1_0:
 *
 * A macro that evaluates to the 1.0 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_0     (G_ENCODE_VERSION (1, 0))

/**
 * CLUTTER_VERSION_1_2:
 *
 * A macro that evaluates to the 1.2 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_2     (G_ENCODE_VERSION (1, 2))

/**
 * CLUTTER_VERSION_1_4:
 *
 * A macro that evaluates to the 1.4 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_4     (G_ENCODE_VERSION (1, 4))

/**
 * CLUTTER_VERSION_1_6:
 *
 * A macro that evaluates to the 1.6 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_6     (G_ENCODE_VERSION (1, 6))

/**
 * CLUTTER_VERSION_1_8:
 *
 * A macro that evaluates to the 1.8 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_8     (G_ENCODE_VERSION (1, 8))

/**
 * CLUTTER_VERSION_1_10:
 *
 * A macro that evaluates to the 1.10 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_10    (G_ENCODE_VERSION (1, 10))

/* evaluates to the current stable version; for development cycles,
 * this means the next stable target
 */
#if (CLUTTER_MINOR_VERSION % 2)
# define CLUTTER_VERSION_CUR_STABLE      (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION + 1))
#else
# define CLUTTER_VERSION_CUR_STABLE      (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION))
#endif

/* evaluates to the previous stable version */
#if (CLUTTER_MINOR_VERSION % 2)
# define CLUTTER_VERSION_PREV_STABLE     (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION - 1))
#else
# define CLUTTER_VERSION_PREV_STABLE     (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION - 2))
#endif

/**
 * CLUTTER_CHECK_VERSION:
 * @major: major version, like 1 in 1.2.3
 * @minor: minor version, like 2 in 1.2.3
 * @micro: micro version, like 3 in 1.2.3
 *
 * Evaluates to %TRUE if the version of the Clutter library is greater
 * than @major, @minor and @micro
 */
#define CLUTTER_CHECK_VERSION(major,minor,micro) \
        (CLUTTER_MAJOR_VERSION > (major) || \
         (CLUTTER_MAJOR_VERSION == (major) && CLUTTER_MINOR_VERSION > (minor)) || \
         (CLUTTER_MAJOR_VERSION == (major) && CLUTTER_MINOR_VERSION == (minor) && CLUTTER_MICRO_VERSION >= (micro)))

/* annotation for exported variables
 *
 * XXX: this has to be defined here because clutter-macro.h imports this
 * header file.
 */
#ifdef _MSC_VER
# ifdef CLUTTER_COMPILATION
#  define CLUTTER_VAR __declspec(dllexport)
# else
#  define CLUTTER_VAR extern __declspec(dllimport)
# endif
#else
# define CLUTTER_VAR extern
#endif

/**
 * clutter_major_version:
 *
 * The major component of the Clutter library version, e.g. 1 if the version
 * is 1.2.3
 *
 * This value can be used for run-time version checks
 *
 * For a compile-time check, use %CLUTTER_MAJOR_VERSION
 *
 * Since: 1.2
 */
CLUTTER_VAR const guint clutter_major_version;

/**
 * clutter_minor_version:
 *
 * The minor component of the Clutter library version, e.g. 2 if the version
 * is 1.2.3
 *
 * This value can be used for run-time version checks
 *
 * For a compile-time check, use %CLUTTER_MINOR_VERSION
 *
 * Since: 1.2
 */
CLUTTER_VAR const guint clutter_minor_version;

/**
 * clutter_micro_version:
 *
 * The micro component of the Clutter library version, e.g. 3 if the version
 * is 1.2.3
 *
 * This value can be used for run-time version checks
 *
 * For a compile-time check, use %CLUTTER_MICRO_VERSION
 *
 * Since: 1.2
 */
CLUTTER_VAR const guint clutter_micro_version;

gboolean clutter_check_version (guint major,
                                guint minor,
                                guint micro);

gboolean clutter_check_windowing_backend (const char *backend_type);

G_END_DECLS

#endif /* __CLUTTER_VERSION_H__ */
